README.md
entro-vault
=================

A CLI to assist environment management using HashiCorp Vault and SSH tunnels for access.

[![oclif](https://img.shields.io/badge/cli-oclif-brightgreen.svg)](https://oclif.io)

<!-- toc -->
* [Getting Started](#getting-started)
* [Usage](#usage)
* [Commands](#commands)
<!-- tocstop -->

# Getting Started
In order to use this tool, you'll first need to install `vault`. Follow the instructions supplied by HashiCorp here [https://developer.hashicorp.com/vault/tutorials/getting-started/getting-started-install#install-vault](https://developer.hashicorp.com/vault/tutorials/getting-started/getting-started-install#install-vault).

You'll also need `ssh` installed and your RSA key has to be added to the server. Please speak to your system administrator if your SSH key has not been added yet.

Finally, a user on Vault needs to be created with you and you need to be associated with the correct policies to gain access to the environment variables you require. Please speak to your system administrator if you have not been added to Vault yet.

## Quick Installation

```bash
npm install -g entro-vault
```

## Full Installation

In order to provide this CLI to all users, we'll want to install it from the binaries provided. Download the `.tar.gz` file from the releases page.

Then run the following to install:
```bash
tar -xvf entro-vault-<version>.tar.gz
sudo rm -rf /usr/local/src/entro-vault
sudo rm -rf /usr/local/bin/entro-vault
sudo mv entro-vault /usr/local/src/entro-vault
sudo ln -s /usr/local/src/entro-vault/bin/entro-vault /usr/local/bin/entro-vault
```

# Usage
<!-- usage -->
```sh-session
$ npm install -g entro-vault
$ entro-vault COMMAND
running command...
$ entro-vault (--version)
entro-vault/1.1.4 linux-x64 node-v16.15.0
$ entro-vault --help [COMMAND]
USAGE
  $ entro-vault COMMAND
...
```
<!-- usagestop -->
# Commands
<!-- commands -->
* [`entro-vault connect`](#entro-vault-connect)
* [`entro-vault disconnect`](#entro-vault-disconnect)
* [`entro-vault env download PATH`](#entro-vault-env-download-path)
* [`entro-vault env upload PATH`](#entro-vault-env-upload-path)
* [`entro-vault help [COMMAND]`](#entro-vault-help-command)
* [`entro-vault wait`](#entro-vault-wait)

## `entro-vault connect`

Connect to the server and begin a reverse tunnel

```
USAGE
  $ entro-vault connect -h <value> -u <value> [-p <value>] [-v <value>] [-l <value>]

FLAGS
  -h, --host=<value>        (required) The IP address or hostname to connect to via SSH
  -l, --listenPort=<value>  [default: 33233] The port to listen on when the tunnel is created
  -p, --port=<value>        [default: 22] The port to connect to via SSH
  -u, --username=<value>    (required) The username to use when connecting via SSH
  -v, --vaultPort=<value>   [default: 8200] The port to connect to the Vault server

DESCRIPTION
  Connect to the server and begin a reverse tunnel

EXAMPLES
  $ entro-vault connect --host=vault.example.com --username=example_user --port=2222
```

_See code: [dist/commands/connect.ts](https://github.com/entrostat/entro-vault/blob/v1.1.4/dist/commands/connect.ts)_

## `entro-vault disconnect`

Disconnect from the server and stop the reverse tunnel

```
USAGE
  $ entro-vault disconnect

DESCRIPTION
  Disconnect from the server and stop the reverse tunnel

EXAMPLES
  $ entro-vault disconnect
```

_See code: [dist/commands/disconnect.ts](https://github.com/entrostat/entro-vault/blob/v1.1.4/dist/commands/disconnect.ts)_

## `entro-vault env download PATH`

Download an environment file from the Vault server

```
USAGE
  $ entro-vault env download [PATH] [-o <value>] [-h <value>]

ARGUMENTS
  PATH  The path to the env that we want to download

FLAGS
  -h, --vault=<value>   [default: http://localhost:33233] The hostname and path of the Vault server
  -o, --output=<value>  [default: ./.env] The path to save the environment file to

DESCRIPTION
  Download an environment file from the Vault server

EXAMPLES
  $ entro-vault env download product:development/backend --output=devops/dev/backend/.env
```

## `entro-vault env upload PATH`

Upload the contents of an environment file to the Vault server

```
USAGE
  $ entro-vault env upload [PATH] -p <value> [-h <value>] [-y]

ARGUMENTS
  PATH  The path to the env that we want to upload

FLAGS
  -h, --vault=<value>        [default: http://localhost:33233] The hostname and path of the Vault server
  -p, --secret-path=<value>  (required) The path to where the env should be stored on Vault
  -y, --yes                  Automatically confirm that the env variables can be overwritten

DESCRIPTION
  Upload the contents of an environment file to the Vault server

EXAMPLES
  $ entro-vault env upload --secret-path=product:development/backend --yes devops/dev/backend/.env
```

## `entro-vault help [COMMAND]`

Display help for entro-vault.

```
USAGE
  $ entro-vault help [COMMAND] [-n]

ARGUMENTS
  COMMAND  Command to show help for.

FLAGS
  -n, --nested-commands  Include all nested commands in the output.

DESCRIPTION
  Display help for entro-vault.
```

_See code: [@oclif/plugin-help](https://github.com/oclif/plugin-help/blob/v5.1.20/src/commands/help.ts)_

## `entro-vault wait`

Wait until the connection has been opened before continuing

```
USAGE
  $ entro-vault wait [-p <value>] [-t <value>]

FLAGS
  -p, --port=<value>     [default: 33233] The port that the tunnel is listening on
  -t, --timeout=<value>  [default: 30] The number of seconds to wait before timing out and returning an error exit code

DESCRIPTION
  Wait until the connection has been opened before continuing

EXAMPLES
  $ entro-vault wait
```

_See code: [dist/commands/wait.ts](https://github.com/entrostat/entro-vault/blob/v1.1.4/dist/commands/wait.ts)_
<!-- commandsstop -->


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [commit-and-tag-version](https://github.com/absolute-version/commit-and-tag-version) for commit guidelines.

### [1.1.4](https://github.com/entrostat/entro-vault/compare/v1.1.3...v1.1.4) (2023-01-01)


### Bug Fixes

* **devops:** Added the readme update to the major release as well ([5c7f662](https://github.com/entrostat/entro-vault/commit/5c7f662e5482fbd18a712fab35af57d8da37108c))

### [1.1.3](https://github.com/entrostat/entro-vault/compare/v1.1.2...v1.1.3) (2023-01-01)


### Bug Fixes

* **devops:** Added the readme script ([616e6b6](https://github.com/entrostat/entro-vault/commit/616e6b64315128014c79aa1c781a2ab204587bb9))

### [1.1.2](https://github.com/entrostat/entro-vault/compare/v1.1.1...v1.1.2) (2023-01-01)


### Bug Fixes

* **deps:** Updated all of the dependencies in the project ([c39a525](https://github.com/entrostat/entro-vault/commit/c39a525d718c884af10026c2d29505f54f59daec))

### [1.1.1](https://github.com/entrostat/entro-vault/compare/v1.1.0...v1.1.1) (2023-01-01)


### Bug Fixes

* **devops:** Updated the target-branch for dependabot ([1969383](https://github.com/entrostat/entro-vault/commit/19693837fd1d26f3905423c77c810d286a6b8342))

## [1.1.0](https://github.com/entrostat/entro-vault/compare/v1.0.2...v1.1.0) (2023-01-01)


### Features

* **wait:** Added the ability to wait for the port to open and then continue, this is cool for CI/CD pipelines ([bb0ea6d](https://github.com/entrostat/entro-vault/commit/bb0ea6d5fbd18a413442f216c45e1f8cf2231973))


### Bug Fixes

* **deps:** bump @oclif/plugin-help from 5.1.19 to 5.1.20 ([acf8381](https://github.com/entrostat/entro-vault/commit/acf8381f2b98ed79800bb4c82758be9d7fdd6bf2))
* **devops:** Updated the version of entro-version ([86a74d1](https://github.com/entrostat/entro-vault/commit/86a74d12514eb02ba15f62c0dddc6b44935c55ff))

### [1.0.2](https://github.com/entrostat/entro-vault/compare/v1.0.1...v1.0.2) (2022-12-04)


### Bug Fixes

* **upload:** Changed the flag to secret-path but didn't use that in the code ([c8957c7](https://github.com/entrostat/entro-vault/commit/c8957c757e06dfbf18486e8f7a2d933f3d6ed12c))

### [1.0.1](https://github.com/entrostat/entro-vault/compare/v1.0.0...v1.0.1) (2022-12-04)


### Bug Fixes

* **devops:** Removed all of the workflows for Github ([df03806](https://github.com/entrostat/entro-vault/commit/df03806fc88cecab558c94a76aba39bf0d7c669b))

## 1.0.0 (2022-12-04)


### Features

* **commands:** Added the ability to disconnect from the ssh command ([dd8fa51](https://github.com/entrostat/entro-vault/commit/dd8fa51b95576a0cfc7173c19a7c3af6e7bf79b3))
* **commands:** Added the ability to download an env file to your machine ([639f477](https://github.com/entrostat/entro-vault/commit/639f477cdd3d071e2d11be124187154e227bc838))
* **commands:** Added the ability to reverse tunnel the Vault port via CLI ([2f12f29](https://github.com/entrostat/entro-vault/commit/2f12f29f6b99385da1d807984c878bc2f0809832))
* **commands:** Added the ability to upload env files ([e2956cf](https://github.com/entrostat/entro-vault/commit/e2956cf4dfba3d38ae4318f0f15103fdaca24106))
* **devops:** Added the entro-version release scripts ([de94fe0](https://github.com/entrostat/entro-vault/commit/de94fe0860a2746b79a49ff5d521204c9cd6f9aa))


test/commands/wait.test.ts
import {expect, test} from '@oclif/test'

describe('wait', () => {
  test
  .stdout()
  .command(['wait'])
  .it('runs hello', ctx => {
    expect(ctx.stdout).to.contain('hello world')
  })

  test
  .stdout()
  .command(['wait', '--name', 'jeff'])
  .it('runs hello --name jeff', ctx => {
    expect(ctx.stdout).to.contain('hello jeff')
  })
})


src/commands/wait.ts
import { Command, Flags } from '@oclif/core';
import * as waitOn from 'wait-on';

export default class Wait extends Command {
    static description =
        'Wait until the connection has been opened before continuing';

    static examples = ['<%= config.bin %> <%= command.id %>'];

    static flags = {
        port: Flags.integer({
            char: 'p',
            description: 'The port that the tunnel is listening on',
            default: 33_233,
        }),
        timeout: Flags.integer({
            char: 't',
            description:
                'The number of seconds to wait before timing out and returning an error exit code',
            default: 30,
        }),
    };

    static args = [];

    public async run(): Promise<void> {
        const { flags } = await this.parse(Wait);
        try {
            await waitOn({
                timeout: flags.timeout * 1000,
                resources: [`tcp:localhost:${flags.port}`],
            });
            this.log(`Connection established!`);
        } catch (e) {
            this.error(
                `Failed to connect to port ${flags.port} after ${flags.timeout} seconds`,
                {
                    code: '1',
                },
            );
        }
    }
}


src/commands/disconnect.ts
import { Command, Flags } from '@oclif/core';
import { getConfig } from '../shared/get-config';
import { saveConfig } from '../shared/save-config';
import { executeCommand } from '../shared/execute-command';

export default class Disconnect extends Command {
    static description =
        'Disconnect from the server and stop the reverse tunnel';

    static examples = ['<%= config.bin %> <%= command.id %>'];

    static flags = {};

    static args = [];

    public async run(): Promise<void> {
        const config = await getConfig(this.config.configDir);
        config.processes = config.processes || [];

        const sshProcessesStr = await executeCommand(
            `ps aux | grep ssh`,
            () => {
                // Do nothing
            },
            console.error,
        );
        const sshProcesses = sshProcessesStr.split('\n');
        const killed: string[] = [];

        for (const command of config.processes) {
            this.log(`Killing process: ${command}`);

            const exists = sshProcesses.find((proc) => proc.includes(command));
            if (!exists) {
                continue;
            }

            const pid = exists.split(/ +/)[1];

            await executeCommand(
                `kill -9 ${pid}`,
                () => {
                    // Do nothing
                },
                console.error,
            );
            killed.push(exists);
        }

        config.processes = config.processes.filter(
            (proc: string) =>
                !killed.includes(proc) && sshProcesses.includes(proc),
        );

        await saveConfig(this.config.configDir, config);
    }
}


src/commands/connect.ts
import { Command, Flags } from '@oclif/core';
import { executeCommand } from '../shared/execute-command';
import { getConfig } from '../shared/get-config';
import { saveConfig } from '../shared/save-config';

export default class Connect extends Command {
    static description = 'Connect to the server and begin a reverse tunnel';

    static examples = [
        `<%= config.bin %> <%= command.id %> --host=vault.example.com --username=example_user --port=2222`,
    ];

    static flags = {
        host: Flags.string({
            char: 'h',
            description: 'The IP address or hostname to connect to via SSH',
            required: true,
        }),
        username: Flags.string({
            char: 'u',
            description: 'The username to use when connecting via SSH',
            required: true,
        }),
        port: Flags.integer({
            char: 'p',
            description: 'The port to connect to via SSH',
            default: 22,
        }),
        vaultPort: Flags.integer({
            char: 'v',
            description: 'The port to connect to the Vault server',
            default: 8200,
        }),
        listenPort: Flags.integer({
            char: 'l',
            description: 'The port to listen on when the tunnel is created',
            default: 33_233,
        }),
    };

    static args = [];

    public async run(): Promise<void> {
        const { flags } = await this.parse(Connect);

        const command = `ssh -L ${flags.listenPort}:localhost:${flags.vaultPort} ${flags.username}@${flags.host} -p ${flags.port} -o StrictHostKeychecking=no`;

        const config = await getConfig(this.config.configDir);
        config.processes = config.processes || [];
        config.processes.push(command);
        await saveConfig(this.config.configDir, config);

        this.log(`


Creating the reverse bind now, when it is complete you can visit:

http://localhost:${flags.listenPort}

To access the Vault server. If you would like to use the Vault CLI, you will need to run the following export command:

export VAULT_ADDR=http://localhost:${flags.listenPort}
        `);

        await executeCommand(command, console.log, console.error);
    }
}
